python/pyspark/sql/types.py

```diff
@@ -619,7 +621,8 @@ def fromInternal(self, obj):
             # it's already converted by pickler
             return obj
         if self._needSerializeAnyField:
-            values = [f.fromInternal(v) for f, v in zip(self.fields, obj)]
+            values = [f.fromInternal(v) if n else v
```

Comment:
Could we run a benchmark with the worst case, when all columns are needed to be converted?  I think here we pay another if and extra element in the zip to prevert function call basically. This one looks okay practically but I guess we should also identify the downside.

Also, let's add a comment here to describe what we are doing here and also add some links to this PR for other guys to refer the benchmarks.

Reply:
I'll add one more optimization here.
And then I'll do benchmarks.

Reply:
Thanks for testing this out. Let's add a comment here to explain what we are doing and move the worst case benchmark into the PR description.